EIPS/eip-2520.md

```diff
@@ -0,0 +1,72 @@
+
+## Specification
+Setting and getting functions **MUST** have the same public interface as specified in EIP 1577. Additionally, they **MUST** also have new public interfaces introduced by this EIP:
+
+* For setting a `contenthash` record, the `setContenthash` **MUST** provide additional `proto` parameter and use it to save the `contenthash`. When `proto` is not provided, it **MUST** save the record as default record.
+
+  ```solidity
+  function setContenthash(bytes32 node, bytes calldata proto, bytes calldata hash) external authorised(node);
+  ```
+
+* For getting a `contenthash` record, the `contenthash` **MUST** provide additional `proto` parameter and use it to get the `contenthash` for requested type. When `proto` is not provided, it **MUST** return the default record.
+
+  ```solidity
+  function contenthash(bytes32 node, bytes calldata proto) external view returns (bytes memory);
+  ```
+
+* Resolver that supports multiple `contenthash` records **MUST** return `true` for `supportsInterface` with interface ID `0x6de03e07`.
+
+Applications that are using ENS `contenthash` records **SHOULD** handle them in the following way:
+
+* If the application only supports one hosting system (like directly handling ENS from IPFS/Swarm gateways), it **SHOULD** request `contenthash` with a specific type. The contract **MUST** then return it and application **SHOULD** correctly handle it.
+
+* If the application supports multiple hosting systems (like MetaMask), it **SHOULD** request `contenthash` without a specific type (like in EIP 1577). The contract **MUST** then return the default `contenthash` record.
+
+## Rationale
+The proposed implementation was chosen because it is simple to implement and supports all important requested features. However, it doesn't support multiple records for the same type and priority order, as they don't give much advantage and are harder to implement properly.
+
+## Backwards Compatibility
+The EIP is backwards-compatible with EIP 1577, the only differences are additional overloaded methods. Old applications will still be able to function correctly, as they will receive the default `contenthash` record.
+
+## Implementation
+```solidity
+contract ContentHashResolver {
+    bytes4 constant private MULTI_CONTENT_HASH_INTERFACE_ID = 0x6de03e07;
+    mapping(bytes32=>mapping(bytes=>bytes)) hashes;
+
+    function setContenthash(bytes32 node, bytes calldata proto, bytes calldata hash) external {
+        hashes[node][proto] = hash;
+        emit ContenthashChanged(node, hash);
+    }
+
+    function contenthash(bytes32 node, bytes calldata proto) external view returns (bytes memory) {
+        return hashes[node][proto];
+    }
+
+    function supportsInterface(bytes4 interfaceID) public pure returns(bool) {
+        return interfaceID == MULTI_CONTENT_HASH_INTERFACE_ID;
+    }
+}
+```
+
+## Copyright
```

Comment:
```suggestion
## Security Considerations
TBD

## Copyright
```
Security considerations section is required.  You can leave it blank for now, and when it comes time to move to Last Call/Final if no issues are brought up you can fill it in with "There are no known security issues." or something.